### ISSUE TITLE
tSQLt framework has Setup

### ISSUE THREAD

**Issue by rapolas-oag** [cagrin/tSQLt-edge#30]
tSQLt has Setup stored procedure which is executed before each test, its quite useful when setting up db for the same test class (schema) but I haven't found this mentioned in your project. Am I right to assume this is not implemented?

```
2. For each of the test cases:
a. A record is created indicating that the test case is being executed in the tSQLt.TestResult table.
b. tSQLt starts a transaction.
c. If there is a stored procedure named SetUp on the test class, it is executed.
d. The test case stored procedure is executed.
e. The transaction is rolled-back.
f. The record in tSQLt.TestResult is updated accordingly if the test case succeeded, failed or threw an error.
```
[Link](https://tsqlt.org/user-guide/test-creation-and-execution/)

**Comment by cagrin** [comment]
Yes, it is not implemented. I also did some changes in detecting what is a test class/method, so be careful.

I am curious, are you going to give a try tSQLt-edge in real work? Please, share your context. What are you going to achieve? I am more than happy to continue this project.

**Comment by rapolas-oag** [comment]
My approach at the moment is more like a POCing tSQLt enablement on existing SQL server databases which have no good/measurable unit test coverage.

I imagined my projects to be ran on Azure Linux agents, and it worked, I like this elegant way of running tSQLt tests in Docker container.

Currently I am a bit concerned regarding tSQLt possible differences within this nuget. Also, JUnit.xml as result is wonderful, but it would be nice to have full coverage report as well.

![image](https://github.com/cagrin/tSQLt-edge/assets/155544794/79058613-0471-48fc-a848-472f83592ce1)

**Comment by cagrin** [comment]
The tSQLt-edge is some kind of POC for MSBuild.Sdk.SqlProj project, especially on Azure SQL Edge. During the implementation I have been testing using "original" unit test from tSQLt and vice-versa. It has been a total fun. I cannot guarantee full compatibility, but to a very large amount of unit tests, it is a solid replica. If you prefer more robust approach, than you can use original tSQLt framework, see /legacy/example.tests. It uses nuget package I have created based on original tSQLt dacpac. It even works with dotnet-sqltest on docker and on windows localhost, but that's another story.

**Comment by rapolas-oag** [comment]
Thank you for the guidance. I think I will jump to legacy approach as it would make sense to setup test sql instance in Docker the way I need for test run.

Could you mention things which are not working with your implementation compared to tsqlt original implementation (not the ones mentioned in documentation)? Like some edge cases you faced and not were able to resolve?

**Comment by rapolas-oag** [comment]
Maybe the last a bit concerning thing when looking into code is ```dotnet add package tSQLt.Dacpacs --version 1.0.8083.3529-preview1``` that its preview version.

**Comment by cagrin** [comment]
> Thank you for the guidance. I think I will jump to legacy approach as it would make sense to setup test sql instance in Docker the way I need for test run.
> 
> Could you mention things which are not working with your implementation compared to tsqlt original implementation (not the ones mentioned in documentation)? Like some edge cases you faced and not were able to resolve?

I think tSQLt-edge is a solid replica of original tSQLt. Even if I initially did it for TDD practice purpose, it ends up to be quite completed. What is missing, besides obvious differences in approach (no CLR, only Azure SQL Database and SQL Server 2019 and 2022 support, etc.):
- no TableToText printing, so AssertEqualTable is saying only "Expected 'table' has different rowset than actual 'table'", no more details
- no SetUp stored procedure support, you have already discovered it
- no tSQLt:NoTransaction flag support
- no tSQLt:SkipTest flag support
- no tSQLt.CaptureOutput, tSQLt.SuppressOutput and other secondary stored procedures.

I think a good strategy for using tSQLt-edge instead of tSQLt is one of two ways:
- start with tSQLt-edge new test project and do no bother original tSQLt,
- start with original tSQLt new test project and after some time try to switch to tSQLt-edge (in theory, it should be easy as changing package name in csproj).

Hope you would give tSQLt-edge a try. I will appreciate any feedback.

**Comment by cagrin** [comment]
> Maybe the last a bit concerning thing when looking into code is `dotnet add package tSQLt.Dacpacs --version 1.0.8083.3529-preview1` that its preview version.

I fixed that and I left only "1.0.8083.3529" version in this nuget package.

**Comment by rapolas-oag** [comment]
All recommendations worked very well, thank you for support.
I jumped to default tSQLt as it is very easy to manage test server the way we need for tSQLt to be in place (which is usually a headache).

Thanks again.
